### ISSUE TITLE
Stacking classmethod and property does not propagate class TypeVar

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#2265]
**Describe the bug**

Annotating the first argument to a function decorated with `@classmethod` and `@property` as a `TypeVar` does not resolve the type from subclasses.

**To Reproduce**

In a venv with pydantic installed, run pyright against the given code.

Revealed types:
- `UserActions[UserT@prisma]`
- `UserActions[UserT@prisma]`

**Expected behavior**

When targeting python3.9 the revealed types should be:
- `UserActions[User]`
- `UserActions[MyUser]`

**Screenshots or Code**

```py
from typing import Generic, TypeVar, Type
from pydantic import BaseModel

UserT = TypeVar('UserT', bound='User')

class UserActions(Generic[UserT]):
    def __init__(self, _model: Type[UserT]) -> None:
        self._model = _model

    def foo(self) -> UserT:
        return self._model(name='Robert')

class User(BaseModel):
    name: str

    @classmethod
    @property
    def prisma(cls: Type[UserT]) -> UserActions[UserT]:
        return UserActions[UserT](cls)

class MyUser(User):
    @property
    def hello(self) -> str:
        return f'Hello, {self.name}!'

if __name__ == '__main__':
    user = User.prisma.foo()
    print(user.name)
    reveal_type(User.prisma)

    custom = MyUser.prisma.foo()
    print(custom.hello)
    reveal_type(MyUser.prisma)
```

**VS Code extension or command-line**
Using the command-line wrapper from [PyPi](https://pypi.org/project/pyright/)

`pyright 1.1.165`

**Additional context**

Mypy correctly infers the return type but does not recognise it as a `property`

```
test.py:27: error: "Callable[[], UserActions[User]]" has no attribute "foo"  [attr-defined]
        user = User.prisma.foo()
               ^
test.py:29: note: Revealed type is 'def () -> test.UserActions[test.User*]'

test.py:31: error: "Callable[[], UserActions[MyUser]]" has no attribute "foo"  [attr-defined]
        custom = MyUser.prisma.foo()
                 ^
test.py:33: note: Revealed type is 'def () -> test.UserActions[test.MyUser*]'
```

Should be noted that the given code will fail at runtime on python3.6-3.8

**Comment by erictraut** [comment]
Thanks for the bug report. I had an incomplete understanding of how classmethod interacted with descriptors. I now understand it, and have improved the modeling of the runtime behavior within the type checker. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.166, which I just published. It will also be included in the next release of pylance.
